Added exercise route fetching and fixed time zones parsing

nikolaydymura opened the pull request.

nikolaydymura committed:
Implemented fetching exercise route information for ExerciseSessionRecord

AliakseiT committed:
Fix Zone Offset Parsing

nikolaydymura committed:
Merge remote-tracking branch 'refs/remotes/quentinleguennec_origin/feature_added_aggregate_function' into feature_added_aggregate_function-merge

# Conflicts:
#	android/src/main/kotlin/dev/duynp/flutter_health_connect/FlutterHealthConnectPlugin.kt
#	lib/src/flutter_health_connect_factory.dart

AliakseiT committed:
Allow records that set HR start/end to the same moment

AliakseiT commented:
@quentinleguennec, your branch seem to be the most advanced of those trying to unblock https://github.com/imdzx/flutter_health_connect, hence we propose our changes to you first.

quentinleguennec commented:
Thanks for the PR @AliakseiT, I'll try to review it this week. 
What was wrong with the time zone parsing exactly?

AliakseiT commented:
> Thanks for the PR @AliakseiT, I'll try to review it this week. What was wrong with the time zone parsing exactly?

Previous code was not able to parse values like "+02:00" using the `map['endZoneOffset'] as int` construct. I believe HC used to send a different value in the map (or it is a vendor-specific field).

quentinleguennec commented on android/src/main/kotlin/dev/duynp/flutter_health_connect/FlutterHealthConnectPlugin.kt:
```diff
@@ -579,6 +597,68 @@ class FlutterHealthConnectPlugin(private var channel: MethodChannel? = null) : F
         }
     }
 
+    private fun onExerciseRoutePermissionCallback(exerciseRoute: ExerciseRoute?) {
```
This is not for a permission is it?
If it is not let's rename it to `onExerciseRouteRequestCallback` for clarity.

quentinleguennec replied:
This code could also be simplified to:
```
        if (exerciseRoute != null) {
            exerciseRecordResult?.put(
                "route", replyMapper.convertValue(
                    exerciseRoute,
                    hashMapOf<String, Any?>()::class.java
                )
            )
        }
        exerciseRouteResult?.success(exerciseRecordResult)
        exerciseRecordResult = null
        exerciseRouteResult = null
```

quentinleguennec commented on android/src/main/kotlin/dev/duynp/flutter_health_connect/FlutterHealthConnectPlugin.kt:
```diff
@@ -579,6 +597,68 @@ class FlutterHealthConnectPlugin(private var channel: MethodChannel? = null) : F
+        exerciseRouteResult = null
+    }
+
+    private fun getRecordById(call: MethodCall, result: Result) {
```
Can you add the following at the start of this function:
```
 if (!checkIfApiReady()) {
            result.error(
                ERROR_NOT_AVAILABLE,
                "The API is not supported.",
                "Maybe the Health Connect app is not installed?"
            )
            return
        }

        if (!::client.isInitialized) {
            client = HealthConnectClient.getOrCreate(context!!)
        }
```

This ensures the API is available and also initialize it if it was not (which can happen if the user just installed the Health app on older devices).

nikolaydymura replied:
Fixed

quentinleguennec commented on android/src/main/kotlin/dev/duynp/flutter_health_connect/FlutterHealthConnectPlugin.kt:
```diff
@@ -579,6 +597,68 @@ class FlutterHealthConnectPlugin(private var channel: MethodChannel? = null) : F
+                        result.success(recordResult)
+                    }
+                } ?: throw Throwable("Unsupported type $type")
+            } catch (e: Throwable) {
```
Can you change the `catch` to this:
```
            } catch (e: SecurityException) {
                result.error(ERROR_MISSING_PERMISSIONS, e.message, e)
            } catch (e: Throwable) {
                result.error(ERROR_UNKNOWN, e.message, e)
            }
```
I'm trying to standardize the error messages.

nikolaydymura replied:
Fixed

quentinleguennec commented on android/src/main/kotlin/dev/duynp/flutter_health_connect/FlutterHealthConnectPlugin.kt:
```diff
@@ -579,6 +597,68 @@ class FlutterHealthConnectPlugin(private var channel: MethodChannel? = null) : F
+                            }
+
+                            is ExerciseRouteResult.NoData -> result.success(recordResult)
+                            else -> result.success(recordResult)
```
This else should never happen, should it? 
If it indeed never should happen then it is best to either throw here, or return an error rather than a success with `result.error(...)`

nikolaydymura replied:
Fixed

quentinleguennec commented on lib/src/flutter_health_connect_factory.dart:
```diff
@@ -295,6 +295,23 @@ class HealthConnectFactory {
+      'type': type.name,
+      'id': id,
+    };
+    final Map<dynamic, dynamic>? data =
```
Can you wrap the channel call and data parsing in this please:
```
 try {
     ...
    } on PlatformException catch (e, stackTrace) {
      throw FlutterHealthConnectException(
        code: e.errorCode,
        message: e.toString(),
        details: e.details,
        stackTrace: stackTrace,
      );
    }
```

This is part of the error standardization.

nikolaydymura replied:
Fixed

quentinleguennec commented on lib/src/utils/datetime_utils.dart:
```diff
@@ -0,0 +1,19 @@
+
+      return Duration(hours: hours * sign, minutes: minutes * sign);
+    }
+    return null;
```
We probably want to throw rather than return null if the parsing fails, else we will never know that something broke if Google Health changes their data structure in the future.
Besides, the code will already throw if  `match.group(2)` or `match.group(3)` are null.

AliakseiT replied:
IMO, throwing is a dangerous tactic with Health Connect - too much of the undocumented variation in how data is serialized. How about we fallback to logging (as in the update to this PR)?

quentinleguennec replied:
TL;DR: Let's do the what you suggested.

Logging errors at the plugin level is usually not a good idea, because it means people using the plugin cannot do their own error logging.

For example, when I use the plugin in production I want to be able to send myself some error reports if something is not working as expected. And since the errors will happen when a user is using the app I will not see any console logs. The only way for me to track and report errors is if the plugin is returning an error.

But your point is valid, if the data is not as expected returning an error means we don't see any of the data at all, even if the only isssue is with the date, which is far from ideal.

I think the only way to handle this cleanly would be to also return an array of warnings for things that went wrong when the plugin sends a result. But I don't have the time to do that now, so let's just do what you suggested.

AliakseiT replied:
Thanks!

`dart:developer log` is a convenient stream abstraction over standard console. To collect local logs from our code and from packages that use `dart:developer log`, we, for instance, use `logger` package wired to the log stream. It get noisy sometimes, though.

```
void setupLogListener(Logger logger) {
  developer.logStream.listen((logRecord) {
    // Convert logRecord to logger levels and forward to logger
    switch (logRecord.level.name) {
      case 'SEVERE':
        logger.e(logRecord.message);
        break;
      case 'WARNING':
        logger.w(logRecord.message);
        break;
      case 'INFO':
        logger.i(logRecord.message);
        break;
      default:
        logger.d(logRecord.message);  // Default to debug level
    }
  });
}

quentinleguennec replied:
That's a cool feature I didn't know about, thanks!

quentinleguennec commented:
@nikolaydymura @AliakseiT Thanks again for the PR!
I added a few comments, most of it is related to some changes I added on top of the original code.
I didn't try to run this PR code yet, but I will after the few points above are fixed.

AliakseiT commented:
> @nikolaydymura @AliakseiT Thanks again for the PR! I added a few comments, most of it is related to some changes I added on top of the original code. I didn't try to run this PR code yet, but I will after the few points above are fixed.

Thanks! We, however, don't see the comments yet. I guess, once you press "request changes" they all get published.

quentinleguennec reviewed (changes requested):
Indeed, I did not press the submit button. Here we go!

nikolaydymura committed:
Disabled empty-serialization error

nikolaydymura committed:
Renamed `onExerciseRoutePermissionCallback` to `onExerciseRouteRequestCallback`

nikolaydymura committed:
Added version check guard

nikolaydymura committed:
Fixed error exceptions

nikolaydymura committed:
Fixed error throwing

nikolaydymura committed:
Wrapped exception on new function

quentinleguennec commented:
@nikolaydymura Thank you for the fixes, let me know when you are happy with the PR and I'll test it on my phone before merging it.

AliakseiT committed:
Add logging and wrap string parsing in try catch

AliakseiT commented:
> @nikolaydymura Thank you for the fixes, let me know when you are happy with the PR and I'll test it on my phone before merging it.

Nikolay is travelling at the moment, let me step in: the PR is ready for the device testing. 

BTW, if you can tell what sources of HC data you experiment with, it helps us to reproduce the findings (if any).

quentinleguennec committed:
Merge branch 'feature_added_aggregate_function' into test_pr

quentinleguennec committed:
[4.1.0]
Fixed issue with timestamps: [#1](https://github.com/quentinleguennec/flutter_health_connect/pull/1) by [AliakseiT](https://github.com/AliakseiT) and [nikolaydymura](https://github.com/nikolaydymura)
Added `getRecordById` [#1](https://github.com/quentinleguennec/flutter_health_connect/pull/1) by [AliakseiT](https://github.com/AliakseiT) and [nikolaydymura](https://github.com/nikolaydymura)
Improved the Example app to add a button for `getRecordById`.

quentinleguennec commented:
@AliakseiT @nikolaydymura I've finally found the time to test the code, and while doing so found that I needed to improve the API a bit to test your work properly.
This is now done and everything seems to be working flawlessly :)
I pushed some commits to this PR to merge the changes I did and add the new function to the Changelog.md.

Feel free to review it and press the merge button if you are happy with the changes.

AliakseiT reviewed (approved):
👍

AliakseiT commented:
Thanks a lot! Looks good. Merge is not available to me (and lets keep it this way :))

quentinleguennec commented:
Thanks again for the PR! :)
